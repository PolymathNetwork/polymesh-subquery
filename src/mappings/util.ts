import { Codec } from "@polkadot/types/types";
import { u8aToString } from "@polkadot/util";
/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);

export const capitalizeFirstLetter = (str: string): string =>
  str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s.replace(/\0/g, "");

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (
  text: string,
  exitOnFirst = false
): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case "(":
      case "{":
      case "<": {
        nestedLevel++;
        break;
      }
      case ")":
      case "}":
      case ">": {
        nestedLevel--;
        break;
      }
      case ",": {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(
      `No top level comma found in ${text}, it probably isn't a map`
    );
  }
  return commas;
};

export const getOrDefault = <K, V>(
  map: Map<K, V>,
  key: K,
  getDefault: () => V
): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};
