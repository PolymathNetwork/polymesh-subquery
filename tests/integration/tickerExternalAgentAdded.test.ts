import { gql } from "@apollo/client/core";
import { getApolloClient } from "../util";
const { query } = getApolloClient();

describe("tickerExternalAgentAdded", () => {
  test("should return the time block and event index when an agent was added to a ticker", async () => {
    const q = {
      variables: { ticker: "12TICKER" },
      query: gql`
        query q($ticker: String!) {
          tickerExternalAgentAddeds(
            filter: {
              ticker: { equalTo: $ticker }
              callerDid: {
                equalTo: "0x0854afeca045161ef42ba9be94c973f65efc0e8532933500865a1fd655148f6c"
              }
            }
          ) {
            nodes {
              callerDid
              datetime
              blockId
              eventIdx
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
  test("should return empty when an agent has been removed", async () => {
    const q = {
      variables: { ticker: "12TICKER" },
      query: gql`
        query q($ticker: String!) {
          tickerExternalAgentAddeds(
            filter: {
              ticker: { equalTo: $ticker }
              callerDid: {
                equalTo: "0xeb686d5391c1123fcd252419bc1e77b525f46bddc8964d34e74be0988c419fea"
              }
            }
          ) {
            nodes {
              datetime
              blockId
              eventIdx
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data?.tickerExternalAgentAddeds.nodes).toEqual([]);
  });
  test("should return empty when the agent is not found", async () => {
    const res = await query({
      variables: { ticker: "12TICKER" },
      query: gql`
        query q($ticker: String!) {
          tickerExternalAgentAddeds(
            filter: {
              ticker: { equalTo: $ticker }
              callerDid: { equalTo: "bogus" }
            }
          ) {
            nodes {
              datetime
              blockId
              eventIdx
            }
          }
        }
      `,
    });

    expect(res?.errors).toBeFalsy();
    expect(res?.data?.tickerExternalAgentAddeds.nodes).toEqual([]);
  });
});
