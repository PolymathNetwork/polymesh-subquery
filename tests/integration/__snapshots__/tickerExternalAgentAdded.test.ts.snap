// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tickerExternalAgentAdded should return the time block and event index when an agent was added to a ticker 1`] = `
Object {
  "tickerExternalAgentAddeds": Object {
    "__typename": "TickerExternalAgentAddedsConnection",
    "nodes": Array [],
  },
}
`;
